### DISCUSSION TITLE
Light Animation while button pressed

### DISCUSSION THREAD

**Discussion by Celerii** [pybricks/support#450] [Q&A]
Hi everyone,

I am discovering pybricks with my LEGO buggy + a PU remote and it is working well, it is so much better than on a smartphone, thanks a lot for this!

Just a small thing that I cannot manage to do with the code: I would like to have the hub's light animated while the right side red button is pressed, but basically only the first color stays active (red, in this case):

```python
    if (Button.LEFT in pressed):
        hub.light.on(Color.BLUE)
    elif (Button.RIGHT in pressed):
        hub.light.animate([Color.RED, Color.ORANGE, Color.YELLOW, Color.GREEN, Color.CYAN, Color.BLUE, Color.VIOLET, Color.MAGENTA], interval=100)
    else:
        hub.light.on(Color.GREEN)
```

I have also tried with the **while(Button.RIGHT in pressed)** but it does not work either... What could be a solution?

Thanks all!

**Comment by laurensvalk** [comment]
Your animation works for me:

```python

from pybricks.hubs import TechnicHub
from pybricks.parameters import Color
from pybricks.tools import wait

hub = TechnicHub()
hub.light.animate([Color.RED, Color.ORANGE, Color.YELLOW, Color.GREEN, Color.CYAN, Color.BLUE, Color.VIOLET, Color.MAGENTA], interval=100)

wait(10*1000)
```

Can you share your whole script instead of just that portion? Could there be another part of your code that's setting it to red?

**Reply by Celerii** [reply]
No problem at all :)

```python
from pybricks.hubs import TechnicHub
from pybricks.pupdevices import Motor, Remote
from pybricks.parameters import Port, Direction, Stop, Button, Color
from pybricks.tools import wait

# Initialize the hub.
hub = TechnicHub()

# Initialize the motors.
steer = Motor(Port.B)
front = Motor(Port.A, Direction.COUNTERCLOCKWISE)

# Connect to the remote.
remote = Remote()

# Read the current settings
old_kp, old_ki, old_kd, _, _ = steer.control.pid()

# Set new values
steer.control.pid(kp=old_kp*4, kd=old_kd*0.4)

# Find the steering endpoint on the left and right.
# The middle is in between.
left_end = steer.run_until_stalled(-200, then=Stop.HOLD)
right_end = steer.run_until_stalled(200, then=Stop.HOLD)

# We are now at the right. Reset this angle to be half the difference.
# That puts zero in the middle.
steer.reset_angle((right_end - left_end)/2)
steer.run_target(speed=200, target_angle=0, wait=False)

# Set steering angle for the buggy
steer_angle = (((right_end - left_end)/2)-5)
print('steer angle:',steer_angle)

# Now we can start driving!
while True:
    # Check which buttons are pressed.
    pressed = remote.buttons.pressed()

    # Choose the steer angle based on the right controls.

    if Button.RIGHT_PLUS in pressed:
        steer.run_target(1400, steer_angle, Stop.HOLD, False)
    elif Button.RIGHT_MINUS in pressed:
        steer.run_target(1400, -steer_angle, Stop.HOLD, False)
    else:
        steer.track_target(0)

    if (Button.LEFT in pressed):
        hub.light.on(Color.BLUE)
    elif (Button.RIGHT in pressed):
        hub.light.animate([Color.RED, Color.ORANGE, Color.YELLOW, Color.GREEN, Color.CYAN, Color.BLUE, Color.VIOLET, Color.MAGENTA], interval=100)
    else:
        hub.light.on(Color.GREEN)

    # Choose the drive speed based on the left controls.
    drive_speed = 0
    if Button.LEFT_PLUS in pressed:
        drive_speed += 100
    if Button.LEFT_MINUS in pressed:
        drive_speed -= 100

    # Apply the selected speed.
    front.dc(drive_speed)

    # Wait.
    wait(10)
```

**Reply by laurensvalk** [reply]
~~Though I haven't seen your code, it's possible that~~ your snippet is part of a loop.

That means you're calling `hub.light.animate` again and again, so you'll only see the first color.

If this is the case, you may want to add some code to start the animation only if it isn't running already.

**Reply by laurensvalk** [reply]
You could do it this way:

At the top of your code somewhere, add:

```python
animation_running = False
```

Then, use it in a way like this:

```python
from pybricks.hubs import TechnicHub
from pybricks.pupdevices import Remote
from pybricks.parameters import Button, Color
from pybricks.tools import wait

hub = TechnicHub()
remote = Remote()

animation_running = False

while True:
    pressed = remote.buttons.pressed()

    if Button.LEFT in pressed:
        hub.light.on(Color.BLUE)
        animation_running = False
    elif Button.RIGHT in pressed and not animation_running:
        hub.light.animate([Color.RED, Color.ORANGE, Color.YELLOW, Color.GREEN, Color.CYAN, Color.BLUE, Color.VIOLET, Color.MAGENTA], interval=100)
        animation_running = True

    wait(10)
```

**Reply by Celerii** [reply]
It's funny, now while pressing the button the light stays... orange 😄

**Reply by laurensvalk** [reply]
I've updated the example above to be self-contained, so it's easier to see a working version in action.

You'll have to expand the logic in your program a bit more to get `animation_running` to behave in the way you want in your bigger program.

In the previous suggestion, there was an `else` statement that always set `animation_running` to false, which is not the desired behavior.

**Reply by Celerii** [reply]
Your last suggestion brings me to the situation I managed to have myself: the animation is triggered when I press the right red button but goes on when I release it. The left red button is the only way to stop it. My goal is to have it last only while the right button is pressed.

Thanks a lot for your help anyway, I do not wish to abuse of your time, this is really for fun on my side, so I hope you do not feel obliged to do all this! I will go on with this by myself: I am just a newbie in Python but I hope to find something.

Thanks again!

**Reply by dlech** [reply]
> My goal is to have it last only while the right button is pressed.

So instead of checking for the left button, how about checking if the right button is not pressed?

```python
    if Button.RIGHT not in pressed and animation_running:
```

**Reply by Celerii** [reply]
Yes @dlech, your solution works, thanks a lot!

For those who would be interested, here is my full script, which makes the right button animate the hub's light while pressed, and set its to red if the left button has been pressed:

```python
    if Button.RIGHT not in pressed and hub_animation_running:
        hub.light.on(Color.GREEN)
        hub_animation_running = False

    if Button.LEFT in pressed:
        hub.light.on(Color.RED)
        hub_animation_running = False
    elif Button.RIGHT in pressed and not hub_animation_running:
        hub.light.animate([Color.RED, Color.ORANGE, Color.YELLOW, Color.GREEN, Color.CYAN, Color.BLUE, Color.VIOLET, Color.MAGENTA], interval=100)
        hub_animation_running = True
```

Thanks again!
